middleware/src/middleware.go

```diff
@@ -322,59 +356,213 @@ func (middleware *Middleware) RestoreHSMSecret() rpcmessages.ErrorResponse {
 	return rpcmessages.ErrorResponse{Success: true}
 }
 
+func (middleware *Middleware) checkMiddlewareSetup() error {
+	passwordSet, err := middleware.redisClient.GetBool(redis.MiddlewarePasswordSet)
+	if err != nil {
+		return err
+	}
+	baseSetupDone, err := middleware.redisClient.GetBool(redis.BaseSetupDone)
+	if err != nil {
+		return err
+	}
+	middleware.isMiddlewarePasswordSet = passwordSet
+	middleware.isBaseSetupDone = baseSetupDone
+	return nil
+}
+
 // UserAuthenticate returns an ErrorResponse struct in response to a rpcserver request.
-// FIXME: This is a dummy implementation of the UserAuthenticate RPC call.
-// in the future this should return an AuthentificationResponse with e.g. an JWT.
-// This currently uses the `dummyIsBaseSetup` boolean, which should be removed when the proper authentication is implemented.
-func (middleware *Middleware) UserAuthenticate(args rpcmessages.UserAuthenticateArgs) rpcmessages.ErrorResponse {
-
-	// TODO: replace the dummyIsBaseSetup with a proper variable loaded from e.g. redis
-	// dummyIsBaseSetup should only be used for the dummy UserAuthenticate RPC and gets reset on middleware restart
-	if !middleware.dummyIsBaseSetup {
-		if args.Username == "admin" && args.Password == "ICanHasPassword?" {
-			// middleware.dummyIsBaseSetup is only set to true after the initial admin password is changed
-			return rpcmessages.ErrorResponse{Success: true}
-		}
-	} else if middleware.dummyIsBaseSetup {
-		dummyUsers := map[string]string{
-			"admin":   middleware.dummyAdminPassword,
-			"satoshi": "shift1",
-			"dev":     "dev",
-		}
-		if expectedPasssword, userIsInMap := dummyUsers[args.Username]; userIsInMap {
-			if args.Password == expectedPasssword {
-				return rpcmessages.ErrorResponse{Success: true}
+// To check if the user should be authenticated from default values, the bool 'isMiddlewarePasswordSet' is read from redis
+func (middleware *Middleware) UserAuthenticate(args rpcmessages.UserAuthenticateArgs) rpcmessages.UserAuthenticateResponse {
+	//Use this structure to hold the bases's user list in memory
+	usersMap := make(map[string]UserAuthStruct)
+
+	// isMiddlewarePasswordSet checks if the base is run the first time.
+	err := middleware.checkMiddlewareSetup()
+	if err != nil {
+		return rpcmessages.UserAuthenticateResponse{
+			ErrorResponse: &rpcmessages.ErrorResponse{
+				Success: false,
+				Message: "authentication failed, redis error",
+				Code:    rpcmessages.ErrorAuthenticationFailed,
```

Comment:
fyi: There is a ErrorCode for a Redis Error called `ErrorRedisError`. But no strong preference on what to use here.

Reply:
I'd prefer to use the authenitcation error here, probably more expresive for the app in this case.